File: Rust/0006-zigzag-conversion.rs
impl Solution {
    pub fn convert(s: String, num_rows: i32) -> String {
        if num_rows == 1 { return s; }
        
        let input: Vec<u8> = s.into_bytes();
        let num_rows = num_rows as usize;

        let mut rows: Vec<String> = vec![String::new(); num_rows];
        let mut current_row: usize = 0;
        let mut direction_up = false;

        for i in 0..input.len() {
            rows[current_row].push(input[i] as char);

            if direction_up {
                if current_row == 0 {
                    direction_up = false;
                    current_row = 1;
                } else {
                    current_row -= 1;
                }
            } else {
                if current_row == num_rows - 1 {
                    direction_up = true;
                    current_row = num_rows - 2;
                } else {
                    current_row += 1;
                }
            }
        }

        return rows.concat();
    }
}

File: c++/0006-zigzag-conversion.cpp
#include <string>

class Solution {
public:
    string convert(string s, int numRows) {        
        if(numRows == 1) { return s; }

        size_t current_row = 0;
        bool direction_up = false;

        vector<string> rows = vector<string>(numRows);
        for(int i = 0; i < numRows; ++ i) 
            rows[i] =  "";

        for(int i = 0; i < s.size(); ++ i) {
            rows[current_row] += s[i];

            if(direction_up) {
                if (current_row == 0) {
                    direction_up = false;
                    current_row = 1;
                } else {
                    current_row -= 1;
                }
            } else {
                if (current_row == numRows - 1) {
                    direction_up = true;
                    current_row = numRows - 2;
                } else {
                    current_row += 1;
                }
            }
        }

        string result = "";

        for(int i = 0; i < numRows; ++ i)
            result += rows[i];

        return result;
    }
};


